### ISSUE TITLE
Enhancement: More selective rule crawling

### ISSUE THREAD

**Issue by alanmcruickshank** [sqlfluff/sqlfluff#3668]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Currently all rules crawl recursively, each themselves evaluating whether they've found the right kind of segment. As an example rule L018 starts with:
```python
# We only trigger on start_bracket (open parenthesis)
if context.segment.is_type("with_compound_statement"):
    ...
```

This is bad for two reasons:
- It's wasteful in performance terms. If we know that a rule will only act on a certain type of segment, then we should be able to crawl much more efficiently and prune avenues which don't contain any of the relevant segments.
- Rules can't influence their own recursion. Any more complex rules which might apply at different levels can't communicate between themselves. This is bad for how I'm thinking of implementing reflow.

I'm going to suggest that rules can declaratively say what kinds of segments they should trigger on, and then that gets taken into account during the crawling to prune areas of the tree that don't contain those segments.

### Use case

Faster linting. More control over recursion.

### Dialect

all

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by pwildenhain** [comment]
Love this -- make a lot of sense. We know many rules will only apply to particular segments, yet they are still *all** run in a query where they may not be used. I think it will make writing new rules much easier too. Happy refactoring!

**Comment by NiallRees** [comment]
Plus 1 this is an opportunity for a big performance improvement. It's not good that every rule has to traverse the tree from scratch, looking for opportunities to make itself useful.

**Comment by barrywhart** [comment]
See these recent issues and PR for some slightly related work on rule performance.

Issues:
* #3037
* #3035

PR #3061 made some progress on the above. In particular, the new class `CrawlBehavior` introduced in the PR is a likely "home" for the new behavior proposed here.
